terraform/modules/mqtt/main.tf

```
@@ -0,0 +1,39 @@
+resource "terraform_data" "instance" {
+  # Store zone since variables not accessible at destroy time.
+  input = var.zone
+  provisioner "local-exec" {
+    command = join(" ", [
+      "gcloud compute instances create-with-container mqtt",
+      "--container-env",
+      join(",", [
+        "MQTT_PASSWORD=${var.mosquitto_password}",
+        "DATABASE_URL=${var.db_conn_str_private}"
+      ]),
+      "--container-image econialabs/inbox:mqtt",
+      "--network ${var.sql_network_id}",
+      "--zone ${var.zone}"
+    ])
+  }
+  provisioner "local-exec" {
+    command = join("\n", [
+      "result=$(gcloud compute instances list --filter NAME=mqtt)",
+      "if [ -n \"$result\" ]; then",
+      join(" ", [
+        "gcloud compute instances delete mqtt",
+        "--quiet",
+        "--zone ${self.output}"
+      ]),
+      "fi"
+    ])
+    when = destroy
+  }
+}
+
+data "external" "ip" {
+  depends_on = [terraform_data.instance]
+  program = [
+    "bash",
+    "-c",
+    "gcloud compute instances list --filter name=mqtt --format 'json(networkInterfaces[0].accessConfigs[0].natIP)' | jq '.[0].networkInterfaces[0].accessConfigs[0]'",
```

```suggestion
    join(" ", [
        "gcloud compute instances list",
        "--filter name=mqtt --format",
        "'json(networkInterfaces[0].accessConfigs[0].natIP)' |",
        "jq '.[0].networkInterfaces[0].accessConfigs[0]'",
    ])
```

- [ ] What happens if runner machine doesn't have `jq`?

---

Well, it's quite hard to manually parse a JSON file. I'd rather specify `jq` as a dep rather than taking the risk (and time) to write a potentially incorrect parser.

---

It looks like the README/init script combo now covers the `jq` dependency, so I agree not worth writing a custom parser

But since there is still another comment outstanding, and for readability/maintainability, how about modifying the command:

```
    join(" ", [
      "gcloud compute instances list",
      "--filter name=mqtt",
      "--format",
      # Parse natIP field from JSON output of gcloud CLI query.
      join(" ", [
        "json(networkInterfaces[0].accessConfigs[0].natIP)",
        "|",
        "jq '.[0].networkInterfaces[0].accessConfigs[0]'"
      ])
    ])
```